---
title: "Summary of npstat pop-gen stats estimates"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

This notebook reads in the raw output from Mathieu Gautier's runs of npstats (from .bam files), 
and saves it as an .RData object. 
It also saves a processed table that is used for plotting and for comparison to other tools.

```{r}
library("here")
library("ggplot2")
library("plyr")
```

```{r}
pops_data<-read.csv(here::here("rdata","dest_v2.samps_8Jun2023.csv"),header=TRUE)
genome_data<-read.csv(here::here("rdata","chroms_file.csv"),header=FALSE)
colnames(genome_data)<-c("chrom","len")
```


```{r}
pops<-pops_data$sampleId
length(pops)
# Remove the SIMULANS "sample"
pops<-pops[grep("SIM",pops,invert = TRUE)]
pop_genome_stats<-data.frame("pop"=rep(pops,each=length(unique(genome_data$chrom[which(genome_data$chrom!="Y")]))))
pop_genome_stats$chrom<-rep(genome_data$chrom[which(genome_data$chrom!="Y")],length(pops))
pop_genome_stats$chrom_len<-rep(genome_data$len[which(genome_data$chrom!="Y")],length(pops))
pop_genome_stats$chrom_good_len<-rep(genome_data$len[which(genome_data$chrom!="Y")],length(pops))
pop_genome_stats$pi<-vector(length=nrow(pop_genome_stats))
pop_genome_stats$med_tajD<-vector(length=nrow(pop_genome_stats))
head(pop_genome_stats)

#chrom<-"2R"
#pop<-"US_Wis_Cro_1_2014-06-25"
files<-list.files(here::here("npstat/06_15/ana_fastp_060823/res_from_bam"))
win<-"50000" # Change this value depending on which window size to read in (10000, 50000, or 100000).
for(pop in pops){
  cat(pop,"\n")
  tab<-data.frame()
  for(chrom in c("2L","2R","3L","3R","4","X")){
    # Total length of chromosome
    chrom_len<-genome_data$len[which(genome_data$chrom==chrom)]
    #tab<-read.table(here::here("rdata/egglib",paste(pop,".stats",sep="")),sep="\t",header=FALSE)
    file<-paste0(pop,".",chrom,".",win,".stats.gz")
    if(file %in% files){
      tmp=read.table(gzfile(here::here("npstat/06_15/ana_fastp_060823/res_from_bam/",file)),header=TRUE)[,1:8]
      tmp$chromosome<-rep(chrom,nrow(tmp))
      tab<-rbind(tab,tmp)
    }
  }
  head(tmp)
  if(nrow(tab)>0){
    tab$pi<-as.numeric(as.character(tab$Pi))
    tab$Tajima_D<-as.numeric(as.character(tab$Tajima_D))
    
    # Summarise PI
    # for each chromosome: Sum the Pi values together, divide by length of chromosome 
    chrom_pi<-tapply((tab$pi*tab$length),
                   INDEX = list(tab$chromosome),sum,na.rm=TRUE)
    pop_genome_stats$pi[which(pop_genome_stats$pop==pop)]<-chrom_pi[
      match(pop_genome_stats$chrom[which(pop_genome_stats$pop==pop)],names(chrom_pi))]
    # Get the total number of good sites for each chromosome
    chrom_good_len<-tapply((tab$length),
                   INDEX = list(tab$chromosome),sum,na.rm=TRUE)
    pop_genome_stats$chrom_good_len[which(pop_genome_stats$pop==pop)]<-chrom_good_len[
      match(pop_genome_stats$chrom[which(pop_genome_stats$pop==pop)],names(chrom_good_len))]
    # Summarise TajD
    # Take the median across all windows for each chromosome,
    chrom_tajD<-tapply(tab$Tajima_D,INDEX = list(tab$chromosome),median,na.rm=TRUE)
    pop_genome_stats$med_tajD[which(pop_genome_stats$pop==pop)]<-chrom_tajD[
      match(pop_genome_stats$chrom[which(pop_genome_stats$pop==pop)],names(chrom_tajD))]
  }
}
head(pop_genome_stats)

save(list=ls(all.names=TRUE),file=here::here("rdata/npstat",paste("npstat_win",win,".RData",sep="")),envir=.GlobalEnv)
write.table(pop_genome_stats,here::here("rdata/npstat",paste("all_pops.win",win,".stats",sep="")),sep="\t",
            col.names=TRUE,row.names = FALSE,quote=FALSE)
```
