---
title: "Summary of EggLib pop-gen stats estimates"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r}
library("here")
library("ggplot2")
library("plyr")
```

```{r}
pops_data<-read.csv(here::here("rdata","dest_v2.samps_8Jun2023_Pass.csv"),header=TRUE)
genome_data<-read.csv(here::here("rdata","chroms_file.csv"),header=FALSE)
colnames(genome_data)<-c("chrom","len")
```

Plotting:
Here I load either the table of all stats or the .RData object.
```{r}
win<-"win10000"
ld<-"rdata"
if(ld=="rdata"){
 load(here::here("rdata/egglib",paste("egglib_",win,".RData",sep="")))
  }else{
  pop_genome_stats<-read.table(here::here("rdata/egglib",paste("all_pops.",win,".stats",sep="")),sep="\t",header=TRUE)
}
pop_genome_stats$pi_sites<-pop_genome_stats$pi/pop_genome_stats$chrom_len
pop_genome_stats$pi_goodsites<-pop_genome_stats$pi/pop_genome_stats$n_goodsites
nrow(pop_genome_stats)
```

Summarise across chromosomes (weighted mean)
Remove X chromosome
```{r}
# Here I remove all data from the X chromosomes 
# and also all data that have undefined pi_goodsites values; these are ones with goodsites = 0 (i.e. no bam files).
pop_genome_stats_summary<-ddply(pop_genome_stats[which(pop_genome_stats$chrom!="X" & is.finite(pop_genome_stats$pi_goodsites)),],
                                .variables = c("pop"),summarise,
                                "pi_wt_mean_chrln"=weighted.mean(pi_sites,chrom_len/sum(chrom_len)),
                                "pi_wt_mean"=weighted.mean(pi_goodsites,chrom_len/sum(chrom_len)),
                                "pi_median"=median(pi_goodsites,na.rm=TRUE),
                                "tajD_wt_mean"=weighted.mean(med_tajD,chrom_len/sum(chrom_len)),
                                "tajD_median"=median(med_tajD,na.rm=TRUE))

# Add continent info to pop_genome_stats_summary
pop_genome_stats_summary$continent<-pops_data$continent[match(pop_genome_stats_summary$pop,pops_data$sampleId)]
head(pop_genome_stats_summary)
length(unique(pop_genome_stats_summary$pop))
pop_genome_stats_summary[which(pop_genome_stats_summary$continent=="Africa"),]

# Remove samples with NA in continent
pop_genome_stats_summary<-pop_genome_stats_summary[which(!is.na(pop_genome_stats_summary$continent)),]
```


```{r}
# Plot pi
ggplot()+
  geom_point(data=pop_genome_stats_summary,
               aes(x=pi_wt_mean,y=pi_wt_mean_chrln,colour=continent))+
  geom_abline(aes(intercept=0,slope=1),linetype="dashed")+
  xlab(expression(paste("Good Sites, ",pi," x 10"^3)))+
  ylab(expression(paste("All Sites, ",pi," x 10"^3)))+
  theme(axis.text=element_text(size=16),
        axis.title=element_text(size=16))

ggplot()+
  geom_point(data=pop_genome_stats_summary,
               aes(x=pi_wt_mean,y=pi_median,colour=continent))+
  geom_abline(aes(intercept=0,slope=1),linetype="dashed")+
  xlab(expression(paste("Good Sites, Wt. Mean, ",pi," x 10"^3)))+
  ylab(expression(paste("All Sites, Median, ",pi," x 10"^3)))+
  theme(axis.text=element_text(size=16),
        axis.title=element_text(size=16))

# Plot pi
ggplot()+
  geom_boxplot(data=pop_genome_stats_summary,
               aes(x=continent,y=pi_wt_mean*1000,fill=continent),width=0.5)+
  geom_point(data=pop_genome_stats_summary,
               aes(x=continent,y=pi_wt_mean*1000,group=continent),position=position_jitter(0.1))+
  xlab("Continent")+
  ylab(expression(paste(pi," x 10"^3)))+
  theme(axis.text=element_text(size=16),
        axis.title=element_text(size=16))

# Plot Tajima's D
ggplot()+
  geom_boxplot(data=pop_genome_stats_summary,
               aes(x=continent,y=tajD_wt_mean),width=0.5)+
  xlab("Continent")+
  ylab("Tajima's D")+
  theme(axis.text=element_text(size=16),
        axis.title=element_text(size=16))

```

