---
title: "Summary of EggLib pop-gen stats estimates"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

This notebook reads in the raw output from my runs of EggLib, and saves it as an .RData object. 
It also saves a processed table that is used for plotting and for comparison to other tools.

```{r}
library("here")
library("ggplot2")
library("plyr")
```

```{r}
pops_data<-read.csv(here::here("rdata","dest_v2.samps_8Jun2023.csv"),header=TRUE)
genome_data<-read.csv(here::here("rdata","chroms_file.csv"),header=FALSE)
colnames(genome_data)<-c("chrom","len")
```


```{r}
pops<-pops_data$sampleId
length(pops)
# Remove the SIMULANS "sample"
pops<-pops[grep("SIM",pops,invert = TRUE)]
pop_genome_stats<-data.frame("pop"=rep(pops,each=length(unique(genome_data$chrom))))
nrow(pop_genome_stats)

pop_genome_stats$chrom<-rep(genome_data$chrom,length(pops))
pop_genome_stats$chrom_len<-rep(genome_data$len,length(pops))
pop_genome_stats$n_goodsites<-vector(length=nrow(pop_genome_stats))
pop_genome_stats$pi<-vector(length=nrow(pop_genome_stats))
pop_genome_stats$med_tajD<-vector(length=nrow(pop_genome_stats))

head(pop_genome_stats)

win<-"win10000"  # Change this value depending on which window size to read in (win10000, win50000, or win100000).
stats_files<-list.files(here::here("rdata/egglib"),pattern = paste(win,".stats",sep=""))
goodsites_files<-list.files(here::here("rdata/egglib"),pattern = paste(win,".goodsites",sep=""))
length(stats_files)
length(goodsites_files)
pop<-pops[1]
for(pop in pops){
  cat(pop,"\n")
  if(any(grepl(pattern = pop,stats_files))){
    if(any(grepl(pop,goodsites_files))){
      stats_tab<-read.table(here::here("rdata/egglib",paste(pop,".",win,".stats",sep="")),sep="\t",header=FALSE)
      colnames(stats_tab)<-c("chromosome", "winstart", "winend", "population", "lseff", "S", "pi", "Tajima_D")
      goodsites_tab<-read.table(here::here("rdata/egglib",paste(pop,".",win,".goodsites",sep="")),sep="\t",
                                header=FALSE)
      colnames(goodsites_tab)<-c("chromosome", "winstart", "winend", "n_good_sites", "prop_good_sites")

      stats_tab$win<-paste(stats_tab$chromosome,":",stats_tab$winstart,"-",stats_tab$winend,sep="")
      goodsites_tab$win<-paste(goodsites_tab$chromosome,":",goodsites_tab$winstart,"-",goodsites_tab$winend,sep="")
      
      #head(stats_tab)
      #head(goodsites_tab)

      stats_tab$win_len<-stats_tab$winend-stats_tab$winstart
      stats_tab$pi<-as.numeric(as.character(stats_tab$pi))
      stats_tab$Tajima_D<-as.numeric(as.character(stats_tab$Tajima_D))

      # Add the number of goodsites to stats tab.
      stats_tab$goodsites<-goodsites_tab$n_good_sites[match(stats_tab$win,goodsites_tab$win)]
      # Compute window-based pisites
      stats_tab$pisites<-stats_tab$pi/stats_tab$goodsites
      # Save a bedgraph file of pi and TajD
      write.table(stats_tab[,which(colnames(stats_tab) %in% c("chromosome","winstart","winend","pisites"))],
                  paste("egglib_browser/",pop,".pi.",win,".bed",sep=""),
                  col.names=FALSE,row.names=FALSE,quote=FALSE,sep="\t")
      write.table(stats_tab[,which(colnames(stats_tab) %in% c("chromosome","winstart","winend","Tajima_D"))],
                  paste("egglib_browser/",pop,".tajD.",win,".bed",sep=""),
                  col.names=FALSE,row.names=FALSE,quote=FALSE,sep="\t")
      
      # Remove windows with < 0.75 of their length covered.
      goodsites_tab<-goodsites_tab[which(goodsites_tab$prop_good_sites>=0.75),]
      stats_tab<-stats_tab[which(stats_tab$win %in% goodsites_tab$win),]
      
      # Summarise goodsites
      # for each chromosome: Sum the number of good sites.
      chrom_goodsites<-tapply(goodsites_tab$n_good_sites,INDEX = goodsites_tab$chromosome,sum,na.rm=TRUE)
      pop_genome_stats$n_goodsites[which(pop_genome_stats$pop==pop)]<-chrom_goodsites
      
      # Summarise PI
      # for each chromosome: Sum the Pi values together.
      chrom_pi<-tapply(stats_tab$pi,INDEX = stats_tab$chromosome,sum,na.rm=TRUE)
      pop_genome_stats$pi[which(pop_genome_stats$pop==pop)]<-chrom_pi

      # Summarise TajD
      # Take the median across all windows for each chromosome,
      chrom_tajD<-tapply(stats_tab$Tajima_D,INDEX = stats_tab$chromosome,median,na.rm=TRUE)
      pop_genome_stats$med_tajD[which(pop_genome_stats$pop==pop)]<-chrom_tajD
    }else{
      cat("No bam file (goodsites data) for: ",pop,"\n")
      stats_tab<-read.table(here::here("rdata/egglib",paste(pop,".",win,".stats",sep="")),sep="\t",header=FALSE)
      colnames(stats_tab)<-c("chromosome", "winstart", "winend", "population", "lseff", "S", "pi", "Tajima_D")

      stats_tab$win<-paste(stats_tab$chromsome,":",stats_tab$winstart,"-",stats_tab$winend,sep="")

      stats_tab$win_len<-stats_tab$winend-stats_tab$winstart
      stats_tab$pi<-as.numeric(as.character(stats_tab$pi))
      stats_tab$Tajima_D<-as.numeric(as.character(stats_tab$Tajima_D))

      # Summarise PI
      # for each chromosome: Sum the Pi values together.
      chrom_pi<-tapply(stats_tab$pi,INDEX = stats_tab$chromosome,sum,na.rm=TRUE)
      pop_genome_stats$pi[which(pop_genome_stats$pop==pop)]<-chrom_pi

      # Summarise TajD
      # Take the median across all windows for each chromosome,
      chrom_tajD<-tapply(stats_tab$Tajima_D,INDEX = stats_tab$chromosome,median,na.rm=TRUE)
      pop_genome_stats$med_tajD[which(pop_genome_stats$pop==pop)]<-chrom_tajD
    }
  }else{
    cat("No stats computed for: ",pop,"\n")
  }
}
nrow(pop_genome_stats)
pop_genome_stats
save(list=ls(all.names=TRUE),file=here::here("rdata/egglib",paste("egglib_",win,".RData",sep="")),envir=.GlobalEnv)
write.table(pop_genome_stats,here::here("rdata/egglib",paste("all_pops.",win,".stats",sep="")),sep="\t",
            col.names=TRUE,row.names = FALSE,quote=FALSE)
```
